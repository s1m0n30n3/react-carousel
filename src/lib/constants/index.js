/**
 * @enum string
 */
export const GalleryMode = {
  BOUNDED: "bounded",
  RETURN: "return",
};

/**
 * @enum string
 */
export const TransitionEffect = {
  SLIDE: "slide",
  FADE: "fade",
  IN_OUT_SLIDE: "inOutSlide",
};

/**
 * @enum string
 */
export const MouseEvent = {
  END: "end",
  START: "start",
  MOVE: "move",
};

/**
 * @enum string
 */
export const CarouselAction = {
  PREV: "prev",
  NEXT: "next",
};

/**
 * @enum string
 */
export const ImageBehavior = {
  FULL: "full",
  CONTAINED: "contained",
};

/**
 * @enum number[]
 */
export const GalleryRatio = {
  R1_1: [1, 1],
  R4_1: [4, 1],
  R3_2: [3, 2],
  R4_3: [4, 3],
  R5_4: [5, 4],
  R16_10: [16, 10],
  R16_9: [16, 9],
  R21_9: [21, 9],
  R32_9: [32, 9],
};

/**
 * @enum string
 */
export const NavigationType = {
  ONLY_ARROWS: "onlyArrows",
  ONLY_DOTS: "onlyDots",
  ARROWS_AND_DOTS: "arrowsAndDots",
  NONE: "none",
};

/**
 * @enum string
 */
export const NavigationPosition = {
  TOP: "top",
  CENTER: "center",
  BOTTOM: "bottom",
  BOTTOM_OUTSIDE: "bottomOutside",
};
