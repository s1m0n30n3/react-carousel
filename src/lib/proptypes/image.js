import PropTypes from "prop-types";

import { imageBehaviorProps, effectProps } from "./common";

export const imageProptypes = {
  imagesLength: PropTypes.number,
  currentIndex: PropTypes.number,
  index: PropTypes.number,
  imageBehavior: PropTypes.oneOf(imageBehaviorProps),
  effect: PropTypes.oneOf(effectProps),
  effectSpeed: PropTypes.number,
  imageBlur: PropTypes.shape({
    blur: PropTypes.bool,
    blurAmount: PropTypes.number,
  }),
  images: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      src: PropTypes.string,
      alt: PropTypes.string,
    })
  ),
};
