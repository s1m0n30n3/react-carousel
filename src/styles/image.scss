.image--base {
    padding: 0;
    margin: 0;
    display: inline-flex;
    height: inherit;
    justify-content: center;
    align-items: center;
    overflow: hidden;
    transform: translate3d(0,0,0);
    img {
        user-select: none;
        object-fit: contain;
    }
}

.wrapper--fade-effect {
    width: max-content;
    overflow: hidden;
}
.wrapper--slide-effect {
    width: 100%;
    overflow: hidden;
    position: relative;
}

/* Effects */
.image--fade-effect {
    transition-timing-function: ease-in-out;
    transition-property: opacity;
    position: absolute;
}

.image--inOutSlide-effect {
    transition-timing-function: ease-in-out;
    transition-property: transform;
}

/* Image Formats */
.image--full-format,
.image--contained-format {
    width: 100%;
    height: 100%;
    img {
        width: inherit;
        height: inherit;
    }
}

.image--full-format img {
    object-fit: cover;
}

.image--contained-format img {
    object-fit: contain;
}