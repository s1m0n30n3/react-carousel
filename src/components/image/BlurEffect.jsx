import React from "react";

import { useBlurEffect } from "hooks";
import { blurEffectPropTypes } from "proptypes/index";

export const BlurEffect = ({ effects, blur, blurAmount, src, ...props }) => {
  const blurProps = useBlurEffect({ blur, blurAmount, src });

  return (
    <div
      {...props}
      className="blur--base"
      style={{ ...effects, ...blurProps }}
    />
  );
};

BlurEffect.propTypes = blurEffectPropTypes;

BlurEffect.displayName = "Gallery.BlurEffect";
