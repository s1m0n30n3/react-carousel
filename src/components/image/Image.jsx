import React from "react";

import { classNames } from "helpers";
import { ImageBehavior, TransitionEffect } from "constants";
import { imageProptypes } from "proptypes/index";
import { useImageTransition } from "hooks";

import { BlurEffect } from "./BlurEffect";

export const Image = ({
  effect,
  index,
  currentIndex,
  images,
  imageBehavior,
  effectSpeed,
  imageBlur,
  ...props
}) => {
  const isFullWidth = imageBehavior === ImageBehavior.FULL;
  const isContained = imageBehavior === ImageBehavior.CONTAINED;
  const isFadeEffect = effect === TransitionEffect.FADE;
  const isInOutSlideEffect = effect === TransitionEffect.IN_OUT_SLIDE;
  const isCurrentSlide = index === currentIndex;

  const effects = useImageTransition({ isCurrentSlide, effectSpeed });

  return (
    <div
      role="listitem"
      className={
        isFadeEffect ? "wrapper--fade-effect" : "wrapper--slide-effect"
      }
    >
      {!isFullWidth && !!imageBlur.blur && (
        <BlurEffect
          effects={effects[effect]}
          src={images[index]?.src}
          {...imageBlur}
        />
      )}
      <figure
        {...props}
        data-cy={`image-slide-${index}`}
        className={classNames([
          "image--base",
          isFadeEffect && "image--fade-effect",
          isFullWidth && "image--full-format",
          isContained && "image--contained-format",
          isInOutSlideEffect && "image--inOutSlide-effect",
          isCurrentSlide && "image--selected",
        ])}
        style={{ ...effects[effect] }}
      />
    </div>
  );
};

Image.propTypes = imageProptypes;

Image.displayName = "Gallery.Image";
