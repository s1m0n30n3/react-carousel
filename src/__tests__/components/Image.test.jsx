import React from "react";
import renderer from "react-test-renderer";
import { describe, expect, test } from "vitest";

import { ImageBehavior, TransitionEffect } from "constants";
import { Image } from "components/image";

describe("Image", () => {
  test("renders correctly without other props, if `imageBlur.blur` is `false`", () => {
    const component = renderer.create(<Image imageBlur={{ blur: false }} />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("renders correctly if `imageBlur.blur` is `true` and images are set", () => {
    const component = renderer.create(
      <Image
        imageBlur={{ blur: true }}
        images={[
          { src: "a.png", alt: "alt image" },
          { src: "b.png", alt: "alt image" },
        ]}
      />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("renders correctly if `imageBlur.blur` is `true` and images are set along with other props", () => {
    const component = renderer.create(
      <Image
        imageBlur={{ blur: true, blurAmount: 20 }}
        effect={TransitionEffect.FADE}
        index={1}
        currentIndex={1}
        imageBehavior={ImageBehavior.CONTAINED}
        effectSpeed={300}
        images={[
          { src: "a.png", alt: "alt image" },
          { src: "b.png", alt: "alt image" },
        ]}
      />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("renders correctly if `imageBehaviour` is `full` and `effect` is `inOutSlide`", () => {
    const component = renderer.create(
      <Image
        imageBlur={{ blur: true, blurAmount: 20 }}
        effect={TransitionEffect.IN_OUT_SLIDE}
        index={1}
        currentIndex={1}
        imageBehavior={ImageBehavior.FULL}
        effectSpeed={300}
        images={[
          { src: "a.png", alt: "alt image" },
          { src: "b.png", alt: "alt image" },
        ]}
      />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
