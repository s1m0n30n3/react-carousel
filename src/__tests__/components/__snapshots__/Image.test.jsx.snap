// Vitest Snapshot v1

exports[`Image > renders correctly if \`imageBehaviour\` is \`full\` and \`effect\` is \`inOutSlide\` 1`] = `
<div
  className="wrapper--slide-effect"
  role="listitem"
>
  <figure
    className="image--base image--full-format image--inOutSlide-effect image--selected"
    data-cy="image-slide-1"
    style={
      {
        "transform": "scale(1)",
        "transitionDuration": "300ms",
      }
    }
  />
</div>
`;

exports[`Image > renders correctly if \`imageBlur.blur\` is \`true\` and images are set 1`] = `
<div
  className="wrapper--slide-effect"
  role="listitem"
>
  <div
    className="blur--base"
    style={
      {
        "backgroundImage": "initial",
        "filter": "blur(undefinedpx)",
        "height": "calc(100% + NaNpx)",
        "left": "-undefinedpx",
        "top": "-undefinedpx",
        "width": "calc(100% + NaNpx)",
        "zIndex": 0,
      }
    }
  />
  <figure
    className="image--base image--selected"
    data-cy="image-slide-undefined"
    style={{}}
  />
</div>
`;

exports[`Image > renders correctly if \`imageBlur.blur\` is \`true\` and images are set along with other props 1`] = `
<div
  className="wrapper--fade-effect"
  role="listitem"
>
  <div
    className="blur--base"
    style={
      {
        "backgroundImage": "url(b.png)",
        "filter": "blur(20px)",
        "height": "calc(100% + 40px)",
        "left": "-20px",
        "opacity": 1,
        "top": "-20px",
        "transitionDuration": "300ms",
        "width": "calc(100% + 40px)",
        "zIndex": 0,
      }
    }
  />
  <figure
    className="image--base image--fade-effect image--contained-format image--selected"
    data-cy="image-slide-1"
    style={
      {
        "opacity": 1,
        "transitionDuration": "300ms",
      }
    }
  />
</div>
`;

exports[`Image > renders correctly without other props, if \`imageBlur.blur\` is \`false\` 1`] = `
<div
  className="wrapper--slide-effect"
  role="listitem"
>
  <figure
    className="image--base image--selected"
    data-cy="image-slide-undefined"
    style={{}}
  />
</div>
`;
